sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -73,132 +79,167 @@
  * <p>The source also accepts an optional configuration: {@code withQueryFilter()} allows you to
  * define a JSON filter to get subset of files in the database.</p>
  *
- * <p>There is also an optional {@code ParseCallback} that can be specified that can be used to
+ * <p>There is also an optional {@code Parser} that can be specified that can be used to
  * parse the InputStream into objects usable with Beam.  By default, MongoDbGridFSIO will parse
  * into Strings, splitting on line breaks and using the uploadDate of the file as the timestamp.
+ * When using a parser that outputs with custom timestamps, you may also need to specify
+ * the allowedTimestampSkew option.
  */
 public class MongoDbGridFSIO {
 
   /**
-   * Function for parsing the GridFSDBFile into objects for the PCollection.
+   * Callback for the parser to use to submit data.
+   */
+  public interface ParserCallback<T> extends Serializable {
+    public void output(T output);
+
+    public void output(T output, @Nullable Instant timestamp);
+  }
+
+  /**
+   * Interface for the parser that is used to parse the GridFSDBFile into
+   * the appropriate types.
    * @param <T>
    */
-  public interface ParseCallback<T> extends Serializable {
-    /**
-     * Each value parsed from the file should be output as an
-     * Iterable of Line&lt;T&gt;.  If timestamp is omitted, it will
-     * use the uploadDate of the GridFSDBFile.
-     */
-    public static class Line<T> {
-      final Instant timestamp;
-      final T value;
-
-      public Line(T value, Instant timestamp) {
-        this.value = value;
-        this.timestamp = timestamp;
-      }
-      public Line(T value) {
-        this.value = value;
-        this.timestamp = null;
-      }
-    };
-    public Iterator<Line<T>> parse(GridFSDBFile input) throws IOException;
+  public interface Parser<T> extends Serializable {
+    public void parse(GridFSDBFile input, ParserCallback<T> callback) throws IOException;
   }
 
   /**
-   * Default implementation for parsing the InputStream to collection of
-   * strings splitting on the cr/lf.
+   * For the default Read&lt;String&gt; case, this is the parser that is used to
+   * split the input file into Strings. It uses the timestamp of the file
+   * for the event timestamp.
    */
-  private static class StringsParseCallback implements ParseCallback<String> {
-    static final StringsParseCallback INSTANCE = new StringsParseCallback();
+  private static class StringParser implements Parser<String> {
```

Comment:
You can also use an anonymous class:
`private static final Parser<String> TEXT_LINES = new Parser<String>() { ... }`